gatereader/reader.py

```diff
@@ -0,0 +1,187 @@
+                raise Reader.CorruptedPacketException()
+            return p, payload
+        except struct.error:
+            # The struct probably cannot be unpacked due to a corrupted
+            # packet
+            raise Reader.CorruptedPacketException()
+
+    def _retransmit_wrapper(self, packet_id, payload, resp_length):
+        """Transmits the packet, receives response while handling Rx errors.
+
+        Attempts to transmit the packet, retransmitting it up to 2 times
+        if RxError is received.
+        After a successful transmission, it attempts to receive the response.
+        If the response is corrupted, it transmits RxError up to 2 times.
+        If this proces fails, this function raises ReaderError.
+        """
+        for i in range(0, 2):
+            try:
+                self._transmit_packet(packet_id, payload)
+                p, payload = self._expect_packet(resp_length)
+                if p.id == PacketId.RX_ERROR:
+                    # The reader does not understand, retransmit
+                    pass
+                else:
+                    return (p, payload)
+            except (Reader.CorruptedPacketException,
+                    serial.SerialTimeoutException):
+                # Corrupted response to our request, request retransmission
+                for j in range(0, 2):
+                    self._transmit_packet(PacketId.RX_ERROR, bytes())
+                    try:
+                        p, payload = self._expect_packet(resp_length)
+                        if p.id == PacketId.RX_ERROR:
+                            # RxError was a response to our RxError. This
+                            # behaviour is unacceptible, let's punish the
+                            # reader by resetting it!
+                            raise Reader.ReaderError()
+                        else:
+                            return (p, payload)
+                    except (Reader.CorruptedPacketException,
+                            serial.SerialTimeoutException):
+                        # Again, corrupted response. Retry
+                        pass
+        # If we got here, we've had 3 consectutive failed receive
+        # attempts. Reader, how dare you! We will reset him for doing
+        # this to us!
+        raise Reader.ReaderError()
+
+    def _check_atr(self):
+        if self.port.inWaiting() == 5:
+            # Trouble is waiting for us in the buffer; there should be
+            # nothing there
+            # HACK ALERT: We must release sooner than we are ready.
+            # This listens for ACK instead of the ATR as the preliminary
+            # version does not have a bootloader
+            try:
+                self._expect_packet(5)
+            except serial.SerialTimeoutException:
+                # Now we are in serious trouble!
+                raise ReaderError()
+            raise ResetException()
+
+    def set_leds(self, leds):
+        self._check_atr()
```

Comment:
`_check_atr()` is not specific for setting LEDs (or beeping or anything) -- put it inside `_retransmit_wrapper` and remove from all these functions.